pkg/apis/kubeone/v1beta1/types.go

```diff
@@ -0,0 +1,331 @@
+	ControlPlane ControlPlaneConfig `json:"controlPlane,omitempty"`
+	// APIEndpoint are pairs of address and port used to communicate with the Kubernetes API
+	APIEndpoint APIEndpoint `json:"apiEndpoint,omitempty"`
+	// CloudProvider configures the cloud provider specific features
+	CloudProvider CloudProviderSpec `json:"cloudProvider,omitempty"`
+	// Versions defines which Kubernetes version will be installed
+	Versions VersionConfig `json:"versions,omitempty"`
+	// ClusterNetwork configures the in-cluster networking
+	ClusterNetwork ClusterNetworkConfig `json:"clusterNetwork,omitempty"`
+	// Proxy configures proxy used while installing Kubernetes and by the Docker daemon
+	Proxy ProxyConfig `json:"proxy,omitempty"`
+	// StaticWorkers describes the worker nodes that are managed by KubeOne/kubeadm
+	StaticWorkers StaticWorkersConfig `json:"staticWorkers,omitempty"`
+	// DynamicWorkers descirbes the worker nodes that are managed by
+	// Kubermatic machine-controller/Cluster-API
+	DynamicWorkers []DynamicWorkerConfig `json:"dynamicWorkers,omitempty"`
+	// MachineController configures the Kubermatic machine-controller component
+	MachineController *MachineControllerConfig `json:"machineController,omitempty"`
+	// Features enables and configures additional cluster features
+	Features Features `json:"features,omitempty"`
+	// Addons are used to deploy additional manifests
+	Addons *Addons `json:"addons,omitempty"`
+	// SystemPackages configure kubeone behaviour regarding OS packages
+	SystemPackages *SystemPackages `json:"systemPackages,omitempty"`
+}
+
+// OperatingSystemName defines the operating system used on instances
+type OperatingSystemName string
+
+var (
+	OperatingSystemNameUbuntu  OperatingSystemName = "ubuntu"
+	OperatingSystemNameCentOS  OperatingSystemName = "centos"
+	OperatingSystemNameCoreOS  OperatingSystemName = "coreos"
+	OperatingSystemNameFlatcar OperatingSystemName = "flatcar"
+	OperatingSystemNameUnknown OperatingSystemName = ""
+)
+
+// HostConfig describes a single control plane node.
+type HostConfig struct {
+	ID                int    `json:"-"`
+	PublicAddress     string `json:"publicAddress"`
+	PrivateAddress    string `json:"privateAddress"`
+	SSHPort           int    `json:"sshPort"`
+	SSHUsername       string `json:"sshUsername"`
+	SSHPrivateKeyFile string `json:"sshPrivateKeyFile"`
+	SSHAgentSocket    string `json:"sshAgentSocket"`
+	Bastion           string `json:"bastion"`
+	BastionPort       int    `json:"bastionPort"`
+	BastionUser       string `json:"bastionUser"`
+	Hostname          string `json:"hostname"`
+	IsLeader          bool   `json:"isLeader"`
+
+	// If not provided (i.e. nil) defaults to TaintEffectNoSchedule, with key
+	// node-role.kubernetes.io/master for control plane nodes.
+	//
+	// Explicitly empty (i.e. []corev1.Taint{}) means no taints will be applied
+	// (this is default for worker nodes).
+	Taints []corev1.Taint `json:"taints,omitempty"`
```

Comment:
@xmudrii WDYT about moving this down to ControlPlaneConfig?

Reply:
For the sake of flexibility, I think it would be best if it stays in the `HostConfig`.

Moving it to `ControlPlaneConfig` would mean that the same taints would be applied to all control plane nodes. That's not always what you want, for example, in HA setups, you can leave one node just for CP components and use other nodes for other workloads as well. This is especially important if machines don't have the same hardware. Same applies for the worker nodes.